### PR TITLE
util-docker,tests: Add compiler tests & Dockerfiles for GCC 14

### PR DESCRIPTION
**Pull request by erin-le**
This commit adds gcc 14 to the compiler tests and Dockerfiles.

### CODE REVIEW

File: .github/workflows/compiler-tests.yaml
Comment (BobbyRBruce) [inline .github/workflows/compiler-tests.yaml:16]: You've accidentally removed `clang-version-15` from the image list.

File: util/dockerfiles/docker-bake.hcl
Comment (BobbyRBruce) [inline util/dockerfiles/docker-bake.hcl:182]: This all looks good to me but did you build the image locally to mature sure this all works? 

To build this image you'd execute `docker buildx bake gcc-version-14` in the "util/dockerfiles" directory. This does require a machine with docker and some other dependencies installed. If you don't have one already get Mahyar to set you up to user docker on amarillo.

If you had permission to push to the GitHub Container registry you'd run `docker build xbake gcc-version-14 --push`, though setting this up a little annoying so I'll do this right now to save you the hassle. This is more just an "FYI" that these steps are necessary.


**Note:** as i had to build this to upload the image, I've verified it builds just fine :)

File: util/dockerfiles/gcc-compiler/Dockerfile
Comment (BobbyRBruce) [inline util/dockerfiles/gcc-compiler/Dockerfile:6]: This is fine, but  just so you know, it's not a requirement that the default is the highest supported. I honestly dosn't matter what the default value is. There probably shouldn't be a defaulk as any user wanting to create image from this should know what version they and be blocked by an error when they don't specify it explicitly... but this isn't particularly important right now.

File: .github/workflows/compiler-tests.yaml
Comment (BobbyRBruce) [inline .github/workflows/compiler-tests.yaml:1]: In a separate commit or PR, can you can look into doing the same for `clang-version-19`? GCC v14 is the latest gcc version and Clang v19 is the last clang version. It's always nice to save we support the latest if possible.

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]
I've uploaded the docker image (ghcr.io/gem5/gcc-version-14:latest). 

I have one fix suggestion in this review that should be addressed.

File: .github/workflows/compiler-tests.yaml
Comment (erin-le) [reply .github/workflows/compiler-tests.yaml:16]: It's still there, it just got moved to the next line by the YAML formatter

Comment (BobbyRBruce) [reply .github/workflows/compiler-tests.yaml:16]: oh i see! Then this PR is perfect then. ! Thanks @erin-le

**Review by BobbyRBruce — APPROVED** [review]

**Comment by erin-le** [comment]
I looked into adding clang-version-19, but since Ubuntu doesn't have a package available for clang-19 yet, I wasn't able to build the image.

**Comment by powerjg** [comment]
When this is committed, I suggest changing the message to "Add dockerfiles for GCC 14" The current commit message sounds like gem5 is getting support for gcc 14, but IIRC, that's not the case right now.

**Review by powerjg — APPROVED** [review]

**Comment by BobbyRBruce** [comment]
> I looked into adding clang-version-19, but since Ubuntu doesn't have a package available for clang-19 yet, I wasn't able to build the image.

That's fine, this is often the case. If you can't install it via apt in 24.04 then let's not bother supporting it.

**Comment by BobbyRBruce** [comment]
> When this is committed, I suggest changing the message to "Add dockerfiles for GCC 14" The current commit message sounds like gem5 is getting support for gcc 14, but IIRC, that's not the case right now.

This PR also adds gcc-14 tests. Doesn't that mean we support GCC 14?

**Review by Harshil2107 — APPROVED** [review]
lgtm

**Comment by powerjg** [comment]
> This PR also adds gcc-14 tests. Doesn't that mean we support GCC 14?

I'd say "no" #1630 

It adds tests, but do they pass?

**Comment by BobbyRBruce** [comment]
> > This PR also adds gcc-14 tests. Doesn't that mean we support GCC 14?
> 
> I'd say "no" #1630
> 
> It adds tests, but do they pass?

Unknown until we merge this PR. 

To stop a chicken and problem arising from this conversation, do you want to rename this PR "util-docker: Add compiler tests & dockerfiles for GCC 14"?

**Pull request merged**
